fairscale/nn/data_parallel/sharded_ddp.py

```diff
@@ -230,21 +230,30 @@ def to(  # type: ignore
         .. note::
             This method modifies the module in-place.
 
+        .. warning:
+            Device changes are not supported, and this will raise an exception. The issue in that case is not
+            really ShardedDDP, but OSS which will not be aware of the device change, and whose buffers will be
+            in a broken state.
+
         Arguments:
             device (:class:`torch.device`): the desired device of the parameters and buffers in this module.
             dtype (:class:`torch.dtype`): the desired floating point type of the floating point parameters and buffers.
             non_blocking (bool): make it an asynchronous call.
 
         Returns:
             Module: self.
-
         """
 
-        for device in self.buckets.keys():
-            for bucket in self.buckets[device]:
+        assert device in self.buckets.keys(), "Changing devices is not supported, because this would break OSSs state"
+        assert (
```

Comment:
changing the tests to make sure that all the combinations were covered (accumulation/amp/manual reduce/buckets) exposed the fact that this was broken, but on the OSS side. OSS remaps the parameters as views of tensors, but it does not see the .to() calls (it's not exposed as a module), so when the module is moved to another device OSS is in a broken state. Asserting here so that this does not go unseen